optuna/samplers/_cmaes.py

```diff
@@ -242,30 +295,46 @@ def sample_relative(
         study._storage.set_trial_system_attr(
             trial._trial_id, "cma:generation", optimizer.generation
         )
+        study._storage.set_trial_system_attr(trial._trial_id, "cma:n_restarts", n_restarts)
         external_values = {
             k: _to_optuna_param(search_space[k], p) for k, p in zip(ordered_keys, params)
         }
         return external_values
 
-    def _restore_or_init_optimizer(
-        self,
-        completed_trials: "List[optuna.trial.FrozenTrial]",
-        search_space: Dict[str, BaseDistribution],
-        ordered_keys: List[str],
-    ) -> CMA:
-
+    def _restore_optimizer(
+        self, completed_trials: "List[optuna.trial.FrozenTrial]",
+    ) -> Tuple[Optional[CMA], int]:
         # Restore a previous CMA object.
         for trial in reversed(completed_trials):
             serialized_optimizer = trial.system_attrs.get(
                 "cma:optimizer", None
             )  # type: Optional[str]
             if serialized_optimizer is None:
                 continue
-            return pickle.loads(bytes.fromhex(serialized_optimizer))
+            n_restarts = trial.system_attrs.get("cma:n_restarts", 0)  # type: int
```

Comment:
I think we can implement IPOP-CMA-ES without introducing any additional system attributes. Although this system attribute, is modified as a counter, it seems like it's only read as a "boolean flag" to determine whether x0 should be initialized uniformly or not, in `_init_optimizer`. The value of this "flag" can be derived from the context when calling `_init_optimizer`, i.e. when called when the optimizer should stop, then initialize x0 uniformly, otherwise, don't. A counter might be useful if we for instance want to cap the number of restarts but I otherwise think we can simplify the implementation.

Reply:
Thank you for your review!
I agree with your opinion. I removed `n_restarts` from system_attrs.